src/crimson/admin/admin_socket.h

```diff
@@ -0,0 +1,306 @@
+using AsokRegistrationRes =
+  std::optional<AdminSocketRef>;  // holding the server alive until after our
+                                  // unregistration
+
+class AdminSocket : public seastar::enable_lw_shared_from_this<AdminSocket> {
+ public:
+  AdminSocket() = default;
+  ~AdminSocket() = default;
+
+  AdminSocket(const AdminSocket&) = delete;
+  AdminSocket& operator=(const AdminSocket&) = delete;
+  AdminSocket(AdminSocket&&) = delete;
+  AdminSocket& operator=(AdminSocket&&) = delete;
+
+  using hook_server_tag = const void*;
+
+  /**
+   *  create the async Seastar thread that handles asok commands arriving
+   *  over the socket.
+   */
+  seastar::future<> init(const std::string& path);
+
+  seastar::future<> stop();
+
+  /**
+   * register an admin socket hooks server
+   *
+   * The server registers a set of APIs under a common hook_server_tag.
+   *
+   * Commands (APIs) are registered under a command string. Incoming
+   * commands are split by spaces and matched against the longest
+   * registered command. For example, if 'foo' and 'foo bar' are
+   * registered, and an incoming command is 'foo bar baz', it is
+   * matched with 'foo bar', while 'foo fud' will match 'foo'.
+   *
+   * The entire incoming command string is passed to the registered
+   * hook.
+   *
+   * \param server_tag  a tag identifying the server registering the hook
+   * \param apis_served a vector of the commands served by this server. Each
+   *        command registration includes its identifying command string, the
+   *        expected call syntax, and some help text.
+   *
+   * A note regarding the help text: if empty, command will not be
+   * included in 'help' output.
+   *
+   * \retval a shared ptr to the asok server itself, or nullopt if
+   *         a block with same tag is already registered.
+   */
+  seastar::future<AsokRegistrationRes> register_server(
+    hook_server_tag server_tag, const std::vector<AsokServiceDef>& apis_served);
+
+  /**
+   * unregister all hooks registered by this hooks-server. The caller
+   * gives up on its shared-ownership of the asok server once the
+   * deregistration is complete.
+   */
+  seastar::future<> unregister_server(hook_server_tag server_tag,
+                                      AdminSocketRef&& server_ref);
+
+ private:
+  /**
+   * the result of analyzing an incoming command, and locating it in
+   * the registered APIs collection.
+   */
+  struct parsed_command_t {
+    std::string m_cmd;
+    cmdmap_t m_parameters;
+    std::string m_format;
+    const AsokServiceDef* m_api;
+    const AdminSocketHook* m_hook;
+    /**
+     *  the length of the whole command-sequence under the 'prefix' header
+     */
+    std::size_t m_cmd_seq_len;
+  };
+  // and the shorthand:
+  using maybe_parsed_t = std::optional<AdminSocket::parsed_command_t>;
+
+  /**
+   * Registering the APIs that are served directly by the admin_socket server.
+   */
+  seastar::future<AsokRegistrationRes> internal_hooks();
```

Comment:
s/internal_hooks/register_admin_hooks/.

Reply:
The name was chosen to distinguish the hooks that are served by the AdminSocket object directly vs. those hooks that are served by the ODS object or the (now defunct) CephContext.

Reply:
thanks for explaining why it is named "internal_hooks". the reasons why i recommend "register_admin_hooks" are

- internal_hooks looks like a noun not an action. and what it returns is not a set of hooks
- as you put, it's not "internal" anymore.

i am not asking for the reason why it's named "internal_hooks".